### PR TITLE
Initial commit of all frontend CFG benchmark generators.

### PR DESCRIPTION
**Pull request by xyzsam**
Includes a basic Makefile to build protos and run unit tests.

### CODE REVIEW

File: frontend/common.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
+"""Common classes for generating benchmarks."""
+
+import cfg_pb2
+
+class IDGenerator(object):
+  """Returns an unused integer as a unique ID."""
+  next_id = 0
+
+  def Next():
```
Comment (rpgolshan) [inline frontend/common.py:9]:
```suggestion
  def Next(self):
```

**Comment by xyzsam** [comment]
Hm, got a email about a comment from you Rob, but I don't see it here anymore. Not sure if you deleted it? Anyways, for that function call I always write `IDGenerator.Next()` (i.e. a static function in a class) rather than invoke it on an instance of `IDGenerator` so it doesn't need `self`.

**Comment by rpgolshan** [comment]
> Hm, got a email about a comment from you Rob, but I don't see it here anymore. Not sure if you deleted it? Anyways, for that function call I always write `IDGenerator.Next()` (i.e. a static function in a class) rather than invoke it on an instance of `IDGenerator` so it doesn't need `self`.

Hey, accidentally posted that too early. I'll have a review up shortly. Your suggestion is good though.

**Review by rpgolshan — COMMENTED** [review]
Can you run flake8 on your code? I will upload a config file in a second but currently I have an edit only for line length:

```
[flake8]
max-line-length = 120
```

Have you thought about adding type hints to your code?

File: frontend/inst_pointer_chase_gen_test.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
+"""Tests for inst_pointer_chase_gen."""
+
+from __future__ import absolute_import  # Not necessary in a Python 3-only module
```
Comment (rpgolshan) [inline frontend/inst_pointer_chase_gen_test.py:3]: Let's support only python3 and delete these lines

File: frontend/inst_pointer_chase_gen_test.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+  def test_generate_callchain_functions(self):
+    self.gen._GenerateCallchainMappings()
+    self.gen._GenerateCallchainFunctions()
+    self.assertEqual(len(self.gen._functions), self.depth * self.num_callchains)
+    # Map from a caller function to its callee in the chain.
+    expected_caller2callee = { 0: 1, 1: 2, 3: 4, 4: 5 }
+
+    # All functions should use the same main code block body.
+    code_block_bodies = set()
+    for func_id, func in self.gen._functions.items():
```
Comment (rpgolshan) [inline frontend/inst_pointer_chase_gen_test.py:51]: This test checks a lot. Can you split this into several functions?

File: frontend/dfs_chase_gen.py [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+
+def register_args(parser):
+  subparser = parser.add_parser(MODULE_NAME)
+  subparser.add_argument('--depth', default=20, type=int,
+                         help='Depth of the function call tree.')
+  subparser.add_argument('--branch_probability', default=0.5, type=float,
+                         help='Branch taken probability.')
+
+
+def _PopRandomFunction(function_list):
```
Comment (rpgolshan) [inline frontend/dfs_chase_gen.py:21]: Duplicated code in inst_pointer_chase. Move to common file?

File: frontend/dfs_chase_gen.py [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+        'int w = z*z + x - y;\n'
+    )
+
+  def _GenerateConditionalBranchCodeBlocks(self, call_targets, probability):
+    if len(call_targets) != 2:
+      raise ValueError(
+          "call_targets must have length 2, got %d" % len(call_targets))
+
+    # Conditional branch taken path.
+    taken_block = self._AddCodeBlock()
```
Comment (rpgolshan) [inline frontend/dfs_chase_gen.py:56]: extract function
```suggestion
    taken_block = self.add_code_block_with_branch(branch_type, targets, probability)
```

File: frontend/common.py [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+    id = IDGenerator.Next()
+    self._code_blocks[id] = cfg_pb2.CodeBlock(id=id)
+    return self._code_blocks[id]
+
+
+  def _FunctionName(self, function_id):
+    return 'function_%d' % function_id
+
+
+  def _AddFunction(self, id):
```
Comment (rpgolshan) [inline frontend/common.py:44]: Rename function to be more descriptive of differences between this and _AddCodeBlock

```suggestion
  def _AddFunctionWithId(self, id):
```

File: frontend/dfs_chase_gen_test.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+  def setUp(self):
+    self.depth = 3
+    self.branch_probability = 0.5
+    self.gen = dfs_chase_gen.DFSChaseGenerator(
+        self.depth, self.branch_probability)
+
+
+  def test_generate_function_tree(self):
+    self.gen._GenerateFunctionTree()
+    self.assertEqual(len(self.gen._function_tree), 2**(self.depth-1) - 1)
```
Comment (rpgolshan) [inline frontend/dfs_chase_gen_test.py:24]: Where do the -1's come from?

File: frontend/inst_pointer_chase_gen.py [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+    # fallthrough branch, and the call, with a return terminator branch.
+
+    for caller, callee in self._caller2callee.items():
+      function = self._AddFunction(caller)
+      main_body = self._AddCodeBlock()
+      main_body.code_block_body_id = self._function_body.id
+      main_body.terminator_branch.type = cfg_pb2.Branch.BranchType.FALLTHROUGH
+      function.instructions.append(main_body)
+
+      if not callee == NO_CALLEE:
```
Comment (rpgolshan) [inline frontend/inst_pointer_chase_gen.py:83]:
```suggestion
      if callee != NO_CALLEE:
```

**Review by rpgolshan — COMMENTED** [review]
I accidentally discarded my other comments. Here are some of the main takeaways

**Review by xyzsam — COMMENTED** [review]
Re type hints: sure, we can add them. Follow up commit?

Re linting - happy to lint. I usually follow google's python style guide, but let's agree on a style set first?

**Comment by rpgolshan** [comment]
> Re type hints: sure, we can add them. Follow up commit?

Sounds good to me.

> Re linting - happy to lint. I usually follow google's python style guide, but let's agree on a style set first?

Yes good idea. Is this Google's python style guide? https://google.github.io/styleguide/pyguide.html
Do you use yapf?

**Comment by xyzsam** [comment]
> > Re linting - happy to lint. I usually follow google's python style guide, but let's agree on a style set first?
> 
> Yes good idea. Is this Google's python style guide? https://google.github.io/styleguide/pyguide.html
> Do you use yapf?

Yes, that's the one. I haven't used `yapf` before but it looks good to me. Formatted everything here with `yapf --style=google --recursive .` Let me know how that looks to you.

**Comment by xyzsam** [comment]
Applied some more fixes from pylint from your PR.

**Comment by xyzsam** [comment]
Renamed all my method names from CamelCase to snake_case.

File: frontend/inst_pointer_chase_gen.py [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+            if callee != NO_CALLEE:
+                call_block = self._AddCodeBlock()
+                # Leave the branch target unspecified for now.
+                call_block.terminator_branch.type = (
+                    cfg_pb2.Branch.BranchType.DIRECT_CALL)
+                call_block.terminator_branch.targets.append(callee)
+                call_block.terminator_branch.taken_probability.append(1)
+                # The end of a function in C will implicitly return, no need to
+                # create another CodeBlock.
+                function.instructions.append(call_block)
```
Comment (rpgolshan) [inline frontend/inst_pointer_chase_gen.py]: Duplicate code in lines 93-97, 69-72

Extract function. Possibly to common class.

File: frontend/inst_pointer_chase_gen_test.py [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+    def test_generate_callchain_function_codeblockbodies(self):
+        self.gen._GenerateCallchainMappings()
+        self.gen._GenerateCallchainFunctions()
+        self.assertEqual(len(self.gen._functions),
+                         self.depth * self.num_callchains)
+
+        # All functions should use the same main code block body.
+        code_block_bodies = set()
+        for func_id, func in self.gen._functions.items():
+            code_block_bodies.add(func.instructions[0].code_block_body_id)
```
Comment (rpgolshan) [inline frontend/inst_pointer_chase_gen_test.py:48]: Split this into a separate loop before line 53. Performance drop is minimal but it is easier to read as it is a separate concept.

File: frontend/inst_pointer_chase_gen_test.py [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+        for func_id, func in self.gen._functions.items():
+            if func_id in expected_caller2callee:
+                # This function calls another in a chain.
+                self.assertEqual(len(func.instructions), 2)
+                self.assertEqual(func.instructions[1].terminator_branch.type,
+                                 cfg_pb2.Branch.BranchType.DIRECT_CALL)
+                expected_callee_func = (
+                    self.gen._functions[expected_caller2callee[func_id]])
+                self.assertEqual(
+                    func.instructions[1].terminator_branch.targets[0],
```
Comment (rpgolshan) [inline frontend/inst_pointer_chase_gen_test.py:75]: These 1s and 0s in the indexes as well as the stream of dots are making this hard to read. Easiest solution I see is replace with a helper function local to this file.
e.g. branch_target_for_func(function)

File: frontend/common.py [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+        """Get the next ID."""
+        IDGenerator.next_id += 1
+        return IDGenerator.next_id
+
+
+def PopRandomElement(somelist):
+    """Pop off a random element from the list."""
+    if not somelist:
+        raise IndexError('PopRandomFunction: list is empty')
+    idx = int(random.random() * len(somelist))
```
[the remark this answers is not in the document]
Comment (xyzsam) [reply frontend/common.py]: Done.

File: frontend/common.py [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+        return 'function_%d' % function_id
+
+    def _AddFunctionWithId(self, next_id):
+        if next_id in self._functions:
+            raise KeyError('there already exists a function with id %d' %
+                           next_id)
+        self._functions[next_id] = cfg_pb2.Function(id=next_id)
+        signature = 'void %s' % self._FunctionName(next_id)
+        sig_body = self._AddCodeBlockBody(signature)
+        self._functions[next_id].signature.code_block_body_id = sig_body.id
```
[the remark this answers is not in the document]
Comment (xyzsam) [reply frontend/common.py]: Done.

File: frontend/dfs_chase_gen.py [hunk 14]
```diff
@@ -0,0 +1,130 @@
+"""Generates an instruction pointer-chase benchmark."""
```
[the remark this answers is not in the document]
Comment (xyzsam) [reply frontend/dfs_chase_gen.py]: Done.

File: frontend/dfs_chase_gen.py [hunk 15]
```diff
@@ -0,0 +1,130 @@
...
+        for caller, callees in self._function_tree.items():
+            for callee in callees:
+                self._AddFunctionWithId(callee)
+            self._functions[caller].instructions.extend(
+                self._GenerateConditionalBranchCodeBlocks(
+                    callees, self._branch_probability))
+
+        for leaf in self._leaf_functions:
+            self._functions[leaf].instructions.append(
+                self._GenerateLeafFunctionCodeBlocks())
```
[the remark this answers is not in the document]
Comment (xyzsam) [reply frontend/dfs_chase_gen.py]: The primary reason for separating this was so I can test the two steps separately.

File: frontend/dfs_chase_gen.py [hunk 16]
```diff
@@ -0,0 +1,130 @@
...
+
+    def GenerateCFG(self):
+        self._GenerateFunctionTree()
+        self._GenerateFunctions()
+        cfg_proto = cfg_pb2.CFG()
+        for func in self._functions.values():
+            cfg_proto.functions.append(func)
+        for cb in self._code_block_bodies.values():
+            cfg_proto.code_block_bodies.append(cb)
+        cfg_proto.entry_point_function = self._root_func
```
[the remark this answers is not in the document]
Comment (xyzsam) [reply frontend/dfs_chase_gen.py]: Done.

File: frontend/inst_pointer_chase_gen.py [hunk 17]
```diff
@@ -0,0 +1,118 @@
+"""Generates an instruction pointer-chase benchmark."""
```
[the remark this answers is not in the document]
Comment (xyzsam) [reply frontend/inst_pointer_chase_gen.py]: Done.

File: frontend/inst_pointer_chase_gen.py (hunk 9 above)
Comment (xyzsam) [reply frontend/inst_pointer_chase_gen.py]: The code looks rather different to me actually. One is for a fallthrough, while the other is for a direct call, which requires a target. Doesn't seem worthwhile to extract.

File: frontend/inst_pointer_chase_gen_test.py (hunk 10 above)
Comment (xyzsam) [reply frontend/inst_pointer_chase_gen_test.py]: Done.

File: frontend/inst_pointer_chase_gen_test.py (hunk 11 above)
Comment (xyzsam) [reply frontend/inst_pointer_chase_gen_test.py]: Done.

File: frontend/dfs_chase_gen.py [hunk 18]
```diff
@@ -0,0 +1,132 @@
...
+        if target:
+            block.terminator_branch.targets.append(target)
+        if probability:
+            block.terminator_branch.taken_probability.append(probability)
+        return block
+
+    def _generate_conditional_branch_code_blocks(self, call_targets,
+                                                 probability):
+        if len(call_targets) != 2:
+            raise ValueError("call_targets must have length 2, got %d" %
```
Comment (rpgolshan) [inline frontend/dfs_chase_gen.py:63]: W1405: Quote delimiter " is inconsistent with the rest of the file

File: frontend/dfs_chase_gen.py [hunk 19]
```diff
@@ -0,0 +1,132 @@
...
+    def generate_cfg(self):
+        self._generate_function_tree()
+        self._generate_functions()
+        return self._generate_cfg(self._functions, self._code_block_bodies,
+                                  self._root_func)
+
+
+def generate_cfg(args):
+    """Generate a CFG of arbitrary callchains."""
+    print("Generating DFS instruction pointer chase benchmark...")
```
Comment (rpgolshan) [inline frontend/dfs_chase_gen.py:130]: W1405: Quote delimiter " is inconsistent with the rest of the file

File: frontend/generate_benchmark.py [hunk 20]
```diff
@@ -0,0 +1,49 @@
...
+            f.write(str(cfg))
+    elif args.output_filename.endswith(".pb"):
+        with open(args.output_filename, "wb") as f:
+            f.write(cfg.SerializeToString())
+    else:
+        raise ValueError("Unknown output file extension %s" %
+                         args.output_filename)
+
+
+if __name__ == '__main__':
```
Comment (rpgolshan) [inline frontend/generate_benchmark.py:48]: W1405: Quote delimiter " is inconsistent with the rest of the file

File: frontend/inst_pointer_chase_gen.py [hunk 21]
```diff
@@ -0,0 +1,120 @@
...
+        self._generate_callchain_mappings()
+        self._generate_callchain_functions()
+        self._generate_entry_function()
+        return self._generate_cfg(self._functions, self._code_block_bodies,
+                                  self._entry_function_id)
+
+
+def generate_cfg(args):
+    """Generate a CFG of arbitrary callchains."""
+    print("Generating instruction pointer chase benchmark...")
```
Comment (rpgolshan) [inline frontend/inst_pointer_chase_gen.py:118]: W1405: Quote delimiter " is inconsistent with the rest of the file

File: frontend/inst_pointer_chase_gen_test.py [hunk 22]
```diff
@@ -0,0 +1,125 @@
...
+        cfg = self.gen.generate_cfg()
+        # +1 accounts for the entry function.
+        num_total_functions = self.depth * self.num_callchains + 1
+        self.assertEqual(len(cfg.functions), num_total_functions)
+        # Each function adds a code block body to store its signature, and all
+        # functions share the same main code block body.
+        self.assertEqual(len(cfg.code_block_bodies), num_total_functions + 1)
+
+
+if __name__ == '__main__':
```
Comment (rpgolshan) [inline frontend/inst_pointer_chase_gen_test.py:124]: W1405: Quote delimiter " is inconsistent with the rest of the file

**Review by rpgolshan — COMMENTED** [review]
Code looks good to me. Marked errors I got from pylint.

**Comment by xyzsam** [comment]
Fixed all quote issues. How are you running pylint? I don't get any of these errors when I run `pylint *.py` (with your pylintc PR in the same directory).

**Comment by rpgolshan** [comment]
> Fixed all quote issues. How are you running pylint? I don't get any of these errors when I run `pylint *.py` (with your pylintc PR in the same directory).

I have a virtual environment with Python 3.8.6+ and pylint installed via pip.
```pylint --rcfile=./pylintrc frontend/*.py```

**Review by rpgolshan — APPROVED** [review]
LGTM. Want to squash and merge as the default pull request merge method?

**Comment by xyzsam** [comment]
Thanks, looks like it was my pylint that was a bit old. Squash and merge SGTM.

**Pull request merged**